libs/vkd3d/acceleration_structure.c

```diff
@@ -0,0 +1,343 @@
+            VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, 0,
+            1, &barrier, 0, NULL, 0, NULL));
+}
+
+static void vkd3d_acceleration_structure_write_postbuild_info(
+        struct d3d12_command_list *list,
+        const D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_DESC *desc,
+        VkDeviceSize desc_offset,
+        VkAccelerationStructureKHR vk_acceleration_structure)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    const struct vkd3d_unique_resource *resource;
+    VkQueryPool vk_query_pool;
+    VkQueryType vk_query_type;
+    uint32_t vk_query_index;
+    VkDeviceSize stride;
+    uint32_t type_index;
+    VkBuffer vk_buffer;
+    uint32_t offset;
+
+    resource = vkd3d_va_map_deref(&list->device->memory_allocator.va_map, desc->DestBuffer);
+    if (!resource)
+    {
+        ERR("Invalid resource.\n");
+        return;
+    }
+
+    vk_buffer = resource->vk_buffer;
+    offset = desc->DestBuffer - resource->va;
+    offset += desc_offset;
+
+    if (desc->InfoType == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_COMPACTED_SIZE)
+    {
+        vk_query_type = VK_QUERY_TYPE_ACCELERATION_STRUCTURE_COMPACTED_SIZE_KHR;
+        type_index = VKD3D_QUERY_TYPE_INDEX_RT_COMPACTED_SIZE;
+        stride = sizeof(uint64_t);
+    }
+    else if (desc->InfoType == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_SERIALIZATION)
+    {
+        vk_query_type = VK_QUERY_TYPE_ACCELERATION_STRUCTURE_SERIALIZATION_SIZE_KHR;
+        type_index = VKD3D_QUERY_TYPE_INDEX_RT_SERIALIZE_SIZE;
+        stride = sizeof(uint64_t);
+        FIXME("NumBottomLevelPointers will always return 0.\n");
+    }
+    else
+    {
+        FIXME("Unsupported InfoType %u.\n", desc->InfoType);
+        /* CURRENT_SIZE is something we cannot query in Vulkan, so
+         * we'll need to keep around a buffer to handle this.
+         * For now, just clear to 0. */
+        VK_CALL(vkCmdFillBuffer(list->vk_command_buffer, vk_buffer, offset,
+                sizeof(uint64_t), 0));
+        return;
+    }
+
+    if (!d3d12_command_allocator_allocate_query_from_type_index(list->allocator,
+            type_index, &vk_query_pool, &vk_query_index))
+    {
+        ERR("Failed to allocate query.\n");
+        return;
+    }
+
+    d3d12_command_list_reset_query(list, vk_query_pool, vk_query_index);
+
+    VK_CALL(vkCmdWriteAccelerationStructuresPropertiesKHR(list->vk_command_buffer,
+            1, &vk_acceleration_structure, vk_query_type, vk_query_pool, vk_query_index));
+    VK_CALL(vkCmdCopyQueryPoolResults(list->vk_command_buffer,
+            vk_query_pool, vk_query_index, 1,
+            vk_buffer, offset, stride,
+            VK_QUERY_RESULT_64_BIT | VK_QUERY_RESULT_WAIT_BIT));
+
+    if (desc->InfoType == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_SERIALIZATION)
+    {
+        /* We'll need some way to store these values for later use and copy them here instead. */
```

Comment:
Is this a TODO? If so, probably should be marked as such.

Reply:
Fixed.